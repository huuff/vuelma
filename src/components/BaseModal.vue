<template>
<div class="modal" :class="{ 'is-active': show }">
  <div 
    class="modal-background"
    @click="tryToCloseByBackdrop"
    ></div>
  <template v-if="$slots.default">
    <div class="modal-content">
      <slot></slot>
    </div>
    <button v-if="closeable"
      class="modal-close is-large"
      aria-label="close"
      @click="$emit('close')"
    ></button>
  </template>
  <template v-if="isCard()">
    <div class="modal-card">
      <header v-if="$slots.cardHeader || cardTitle || closeable" class="modal-card-head">
        <slot name="cardHeader"></slot>
        <p v-if="cardTitle" class="modal-card-title">
          {{ cardTitle }}
        </p>
        <button v-if="closeable"
          type="button"
          class="delete"
          aria-label="close"
          @click="$emit('close')"
        ></button>
      </header>
      <section v-if="$slots.cardBody" class="modal-card-body">
        <slot name="cardBody"></slot>
      </section>
      <footer v-if="$slots.cardFooter" class="modal-card-foot">
        <slot name="cardFooter"></slot>
      </footer>
    </div>
  </template>
</div>
</template>

<script setup lang="ts">
import { useSlots, computed } from 'vue';

const props = withDefaults(defineProps<{
  show?: boolean;
  closeable?: boolean;
  backdropCloseable?: boolean;
  cardTitle?: string;
}>(), {
  show: true,
  closeable: true,
  backdropCloseable: undefined,
});

// Sets a sensible default for backdropCloseable,
// if no value is provided, it's assumed to be
// backdropCloseable if it's closeable. Otherwise,
// we just use the provided prop
const computedBackdropCloseable = computed(() => {
  if (props.backdropCloseable === undefined) {
    return props.closeable;
  } else {
    return props.backdropCloseable;
  }
});

const emit = defineEmits<{
  (event: 'close'): void
}>();

const slots = useSlots();
function isCard() {
  return slots.cardHeader || slots.cardFooter || slots.cardBody || props.cardTitle;
}

function tryToCloseByBackdrop() {
  if (computedBackdropCloseable.value) {
    emit('close');
  }
}

if (slots.default && isCard()) {
  throw new Error('A modal can only be normal (only default slot is filled), or a card (any of cardHeader, cardBody or cardFooter filled)');
}

if (slots.cardHeader && props.cardTitle) {
  throw new Error("Can't have both a cardTitle and a cardHeader in a modal!");
}
</script>
