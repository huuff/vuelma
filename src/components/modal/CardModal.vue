<template>
  <div class="modal-card">
    <header v-if="$slots.header || title || closeable" class="modal-card-head">
      <slot name="cardHeader"></slot>
      <p v-if="title" class="modal-card-title">
        {{ title }}
      </p>
      <button v-if="closeable"
        type="button"
        class="delete"
        aria-label="close"
        @click="emit('close')"
      ></button>
    </header>
    <section v-if="$slots.default" class="modal-card-body">
      <slot></slot>
    </section>
    <footer v-if="$slots.footer" class="modal-card-foot">
      <slot name="footer"></slot>
    </footer>
  </div>
</template>

<script setup lang="ts">
import { useSlots } from 'vue';

const props = withDefaults(defineProps<{
  closeable?: boolean;
  title?: string;
}>(), {
  closeable: true,
});

const emit = defineEmits<{
  (event: 'close'): void
}>();

const slots = useSlots();
if (slots.header && props.title) {
  throw new Error("Can't have both a title and a header in a CardModal!");
}
</script>
