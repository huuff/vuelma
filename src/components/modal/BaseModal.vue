<template>
<div class="modal" :class="{ 'is-active': actualShow }">
  <div 
    class="modal-background"
    @click="tryToCloseByBackdrop"
    ></div>

  <render></render>

  <button v-if="closeable && !isCard()"
    class="modal-close is-large"
    aria-label="close"
    @click="actualShow = false"
  ></button>
</div>
</template>

<script setup lang="tsx">
import { useSlots, computed, toRef, cloneVNode } from 'vue';
import { useOptionalTwoWayBinding } from '@/composables/optional-two-way-binding';
import CardModal from './CardModal.vue';
import partial from "lodash/partial";

const props = withDefaults(defineProps<{
  show?: boolean;
  closeable?: boolean;
  backdropCloseable?: boolean;
}>(), {
  show: undefined,
  closeable: true,
  backdropCloseable: undefined,
});

const emit = defineEmits<{
  (event: 'update:show', value: boolean): void
}>();

const actualShow = useOptionalTwoWayBinding(true, toRef(props, 'show'), partial(emit, "update:show"));

// Sets a sensible default for backdropCloseable,
// if no value is provided, it's assumed to be
// backdropCloseable if it's closeable. Otherwise,
// we just use the provided prop
const computedBackdropCloseable = computed(() => {
  if (props.backdropCloseable === undefined) {
    return props.closeable;
  } else {
    return props.backdropCloseable;
  }
});


const slots = useSlots();
function isCard() {
  if (slots.default) {
    const defaultSlot = slots.default();
    return defaultSlot.length === 1 && defaultSlot[0].type === CardModal;
  } else {
    return false;
  }
}

function tryToCloseByBackdrop() {
  if (computedBackdropCloseable.value) {
    emit('update:show', false);
  }
}

const render = () => {
  if (isCard())
    return slots.default && cloneVNode(slots.default()[0], {
      onClose: () => actualShow.value = false,
      ":closeable": toRef(props, "closeable")
    });
  else
    return <div class="modal-content"> { slots.default && slots.default() } </div>
}
</script>
