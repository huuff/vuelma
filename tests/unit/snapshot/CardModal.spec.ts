import { mount } from "@vue/test-utils";
import BaseModal from "@/components/modal/BaseModal.vue";
import CardModal from "@/components/modal/CardModal.vue";

test("CardModal.vue snapshot", () => {
  const wrapper = mount(BaseModal, {
    props: {
      show: true,
      closeable: true,
    }, 
    global: {
      components: { CardModal },
    },
    slots: {
      default: `
        <card-modal titleText="Card Modal Title">
          <template #default>
            <p>Card modal body</p>
          </template>
          <template #footer>
            <p>Card modal footer</p>
          </template>
        </card-modal>
      `,
    },
  });

  expect(wrapper.element).toMatchSnapshot();
});
