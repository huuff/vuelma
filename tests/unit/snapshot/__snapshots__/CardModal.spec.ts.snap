// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CardModal.vue snapshot 1`] = `
<div
  class="modal is-active"
>
  <div
    class="modal-background"
  />
  <div
    class="modal-card"
  >
    <header
      class="modal-card-head"
    >
      
      
      <p
        class="modal-card-title"
      >
        Card Modal Title
      </p>
      <button
        aria-label="close"
        class="delete"
        type="button"
      />
    </header>
    <section
      class="modal-card-body"
    >
      
      <p>
        Card modal body
      </p>
      
    </section>
    <footer
      class="modal-card-foot"
    >
      
      <p>
        Card modal footer
      </p>
      
    </footer>
  </div>
  <!--v-if-->
</div>
`;
